Improve non-core contract CGP decoding with proxies

bowd opened the pull request:
### Description

This is a follow-up PR to #10019, which improved the tooling around transactions targeting non-core contracts in `celocli` by using Sourcify to query for metadata. While testing the upcoming Mento upgrade (MU01), we realized that two pieces of missing functionality around proxy initialization are stopping us from decoding the proposal. 

a. Decoding the proxy `_setAndInitializeImplementation` method's nested call to `initialize` on the implementation. This works for core contracts but needed a bit of love to make it work for external contracts that use the same Proxy implementation.

```yaml
proposal: 
  0: 
    contract: BreakerBoxProxy(0x9268466581dbd4f0d0231f471C7dCf7e622EB5f8)
    function: _setAndInitializeImplementation
    args: 
      0: 0x30D969B9df1FbD766ba15ce1ceEDFffD5EA4B3F7
      1: 0x462d0b2e0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000fdd8bd58115ffbf04e47411c1d228ecc45e930750000000000000000000000000000000000000000000000000000000000000004000000000000000000000000874069fa1eb16d44d622f2e0ca25eea172369bc100000000000000000000000010c892a6ec43a53e45d0b916b4b7d383b1b78c0f000000000000000000000000e4d517785d091d3c54818832db6094bcc2744545000000000000000000000000a1a8003936862e7a15092a91898d69fa8bce290c
    params: 
      __length__: 2
      implementation: 0x30D969B9df1FbD766ba15ce1ceEDFffD5EA4B3F7
      callbackData: 0x462d0b2e0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000fdd8bd58115ffbf04e47411c1d228ecc45e930750000000000000000000000000000000000000000000000000000000000000004000000000000000000000000874069fa1eb16d44d622f2e0ca25eea172369bc100000000000000000000000010c892a6ec43a53e45d0b916b4b7d383b1b78c0f000000000000000000000000e4d517785d091d3c54818832db6094bcc2744545000000000000000000000000a1a8003936862e7a15092a91898d69fa8bce290c
      initialize@462d0b2e: 
        __length__: 2
        _rateFeedIDs: 
          0: 0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1
          1: 0x10c892A6EC43a53E45D0B916B4b7D383B1b78C0F
          2: 0xE4D517785D091D3c54818832dB6094bcc2744545
          3: 0xA1A8003936862E7a15092A91898D69fa8bCE290c
        _sortedOracles: 0xFdd8bD58115FfBf04e47411c1d228eCC45E93075
    value: 0
```

Example output from this branch which shows that the call to the external BreakerBoxProxy also gets the `callbackData` decoded using the BreakerBox initialize method's abi.

b. Keeping track of proxy implementations set during the proposal. The tooling already understands external contract proxies but only if the implementation is set and can be read on-chain. However, if the first implementation is set during the proposal, it will fail to decode subsequent calls on the proxy that actually target the implementation.

In order to fix this I added a `proxyImplementationOverride` mapping to the `block-explorer` package, which gets updated while parsing the proposal to keep track of the most recent implementation address. So when querying for the ABI, it will refer to that instead. 

### Other changes

- In order to achieve the above I also ended up refactoring some of the code to make it cleaner, specifically, in the `block-explorer` we now build a `ContractDetails` mapping either from core contracts or from sourcify and use that as a synergy point for the rest of the code, instead of the multiple branches that existed before.
- I also removed some of the special case code introduced in the first iteration in #9735, which were no longer required

### Tested

Tests were updated

### Related issues

Fixes - [#178](https://github.com/mento-protocol/mento-general/issues/178)

### Backwards compatibility

All external methods were kept.

### Documentation

N/A

bowd committed:
fix: identify the proxy bug and implement a fix

bowd committed:
refactor: move to using ConractMapping as the central point

bowd committed:
refactor: remove try-cache in favour of an explicit check

bowd committed:
fix: remove too detailed debugs

bowd committed:
fix: remove too detailed debugs

ninabarbakadze reviewed (commented):
let's make the ci happy @bowd

ninabarbakadze committed:
test: fix the test that expected null instead of undefined

ninabarbakadze committed:
style: make linter happy

ninabarbakadze committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

ninabarbakadze committed:
fix: fix handle promises appropriately

ninabarbakadze committed:
Merge branch 'bowd/fix-proxy-bug-in-cgp-decoding' of https://github.com/celo-org/celo-monorepo into bowd/fix-proxy-bug-in-cgp-decoding

ninabarbakadze committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

bowd committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

bowd committed:
fix: refactor to optimise

bowd commented:
@aaronmgdr / @dckesler pinging you as you've been involved in the last iteration of this work. I've made a few improvements to the external decoding around proxies, which are needed in order for the upcoming CGP that we want to publish. 
I also think I made the block-explorer code a bit more readable now, by focusing everything around the "ContractMapping" type. 

Would love your thoughts on this 🙏

bowd committed:
refactor: slight tweaking

bowd committed:
refactor: undo rename, reduce diff

bowd committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

bowd committed:
fix: make submission from json work

bowd committed:
fix: lint issues

bowd committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

dckesler commented on packages/sdk/explorer/src/block-explorer.ts:
```diff
@@ -111,7 +105,7 @@ export class BlockExplorer {
   }
 
-  async tryParseTx(tx: CeloTxPending): Promise<null | ParsedTx> {
+  async tryParseTx(tx: CeloTxPending): Promise<ParsedTx | null> {
```
Why did these get flipped?

bowd replied:
I guess I just like having the null/undefined at the tail of the type union, more of a stylistic thing but happy to revert if needed.

aaronmgdr replied:
i think its good to have desired first

dckesler replied:
I just wanted to know if I was missing something. I agree it looks nicer now.

dckesler commented on packages/sdk/explorer/src/block-explorer.ts:
```diff
@@ -74,9 +61,16 @@ export class BlockExplorer {
-  async updateContractDetailsMapping(name: CeloContract, address: string) {
-    const cd = await getContractDetailsFromContract(this.kit, name, address)
-    this.addressMapping.set(cd.address, getContractMappingFromDetails(cd))
+  async updateContractDetailsMapping(name: string, address: string) {
```
We should keep `name` as a `CeloContract` if possible.

dckesler commented on packages/sdk/explorer/src/block-explorer.ts:
```diff
@@ -123,106 +117,18 @@ export class BlockExplorer {
     }
   }
 
-  getContractMethodAbiFromCore = (
```
Why are we deleting these methods? Would it be okay if we left them?

dckesler replied:
It looks like we're still deleting these methods which would still be a breaking change. Were you going to add these back?

aaronmgdr commented on packages/sdk/explorer/src/block-explorer.ts:
```diff
@@ -74,9 +61,16 @@ export class BlockExplorer {
-    this.addressMapping.set(cd.address, getContractMappingFromDetails(cd))
+  async updateContractDetailsMapping(name: string, address: string) {
+    if (isCoreContract(name)) {
+      const cd = await getContractDetailsFromContract(this.kit, name, address)
```
nit `cd` => `contractDetails`

dckesler replied:
+1

bowd replied:
👍

aaronmgdr commented on packages/sdk/explorer/src/sourcify.test.ts:
```diff
@@ -7,12 +7,12 @@ import {
 const CONTRACT_METADATA = require('../fixtures/contract.metadata.json')
 // This is taken from protocol/contracts/build/AccountProxy.json
-const PROXY_METADATA = require('../fixtures/proxy.metadata.json')
+// const PROXY_METADATA = require('../fixtures/proxy.metadata.json')
```
consider removing over commenting out

aaronmgdr commented on packages/sdk/governance/src/proposals.ts:
```diff
@@ -298,37 +321,58 @@ export class ProposalBuilder {
+    // let to: string = tx.contract
+    // if (metadata && metadata.contractName) {
+    //   to = tx.contract
+    // }
```
remove?

aaronmgdr reviewed (commented):
My suggestion would be a few typedoc comments over the main public functions

dckesler reviewed (commented):
The BlockExplorer class is made available to anyone using the `@celo/explorer` package which includes the methods you removed or changed. So ultimately this would be quite a few breaking changes. If we need it to be breaking changes that's fine and we can work on it, but in the description it said there were none so I wanted to bring this up first.

bowd committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding

bowd committed:
fix: fix review comments

bowd committed:
fix: restore deleted methods

bowd committed:
refactor: add comments

bowd commented:
> The BlockExplorer class is made available to anyone using the @celo/explorer package which includes the methods you removed or changed. So ultimately this would be quite a few breaking changes. If we need it to be breaking changes that's fine and we can work on it, but in the description it said there were none so I wanted to bring this up first.

@dckesler you're absolutely right. I was short-sighted I thought that it was ok because those methods aren't actually used anywhere in celo-monorepo anymore but that doesn't mean anything. I should have made some of those methods private in the previous PRs but that's also on me, in order to truly make things backward compatible I brought everything back.

Also added some typedoc (cc @aaronmgdr) and fixed the other suggestions 👍.

Would it be possible to get a release going for the cli to be able to get these changes live? We're planning to release on Monday and approvers would need this functionality in the CLI to be able to verify the proposal, is that feasible on your end?

bowd committed:
fix: treat case when a proposal transaction references a core contract but with a different address

aaronmgdr commented on packages/sdk/governance/src/proposals.ts:
```diff
@@ -139,62 +140,82 @@ export const proposalToJSON = async (
       value: parsedTx.tx.value,
     }
 
+    debug(jsonTx)
```
intended for prod release?

bowd committed:
fix: add missing function

bowd committed:
fix: remove debug

bowd committed:
fix: add getContractMethodAbiFallback and mark all as deprecated

dckesler committed:
Merge branch 'master' into bowd/fix-proxy-bug-in-cgp-decoding
